### PR TITLE
First pass at Shallow Gravy support

### PR DESCRIPTION
**Pull request by binford2k**
This should now theoretically support both Shallow Gravy and 3.8.x.
There were two main avenues for porting to the new filesystem layout.
First I parameterized `$codedir` and `$confdir` where I reasonably
could, so that if there are changes in the future, it should be
relatively easy to port. Second, I wrote a `classroom::compatibility`
shim class that drops in some janky symlinks so that legacy PE
filesystem layouts will work unchanged with the new layout. This means
that when it comes time to drop 3.x support, we just drop the shim.

This isn't 100% parameterized, because some of the paths exist in static
files, but the paths that would need corrected only exist in three
files:

* `files/pre-commit`
* `files/process_reports.rb`
* `files/sudoers.classroom`

I ended up conflating a lot of overall cleanups and code refactoring
into this, simply because I was touching the same lines of code and I
didn't want to write and rewrite the code multiple times.

I should also note that this isn't truly complete, as the `pltraining/rbac`
and other modules may need updated.

### CODE REVIEW

File: manifests/params.pp [hunk 1]
```diff
@@ -38,7 +40,10 @@
   $time_servers = ['0.pool.ntp.org iburst', '1.pool.ntp.org iburst', '2.pool.ntp.org iburst', '3.pool.ntp.org']
 
   # where the agent installer tarball should go. This is only relevant when promoting a secondary master
-  $publicdir = '/opt/puppet/packages/public/classroom'
+  $publicdir = $aio ? {
```
Comment (joshsamuelson) [inline manifests/params.pp]: I'm getting an error here, it looks like $aio isn't a thing?

Comment (binford2k) [inline manifests/params.pp]: bah, that was a variable I calculated, but then realized that I could just use `aio_agent_version`

File: manifests/agent.pp [hunk 2]
```diff
@@ -4,35 +4,17 @@
...
 #  * hiera configuration
 #  * time synchronization with the classroom master
-class classroom::agent (
-  $workdir     = $classroom::workdir,
-  $autosetup   = $classroom::autosetup,
-  $managerepos = $classroom::managerepos,
-  $password    = $classroom::password,
-  $consolepw   = $classroom::consolepw,
-) inherits classroom {
+class classroom::agent {
```
Comment (joshsamuelson) [inline manifests/agent.pp]: Doesn't this need to inherit?

**Comment by binford2k** [comment]
No. Because I'm ensuring that it can never be called on it's own, we can assume that the `classroom` class has been included, so its variables are in the catalog.

**Comment by binford2k** [comment]
I'm trying to slowly migrate it back towards best practices instead of overly clever.

File: manifests/params.pp [hunk 3]
```diff
@@ -8,17 +8,16 @@
...
 
-  # list of classes that should be available in the console
-  $classes   = [ 'users', 'apache', 'userprefs' ]
-
   if $::osfamily == 'windows' {
     # Path to the student's working directory
     $workdir = 'c:/puppetcode'
-    $etcpath = 'C:/ProgramData/PuppetLabs/puppet/etc'
+    $confdir = 'C:/ProgramData/PuppetLabs/puppet/etc'
+    $codedir = 'C:/ProgramData/PuppetLabs/puppet/etc'      # TODO
```
Comment (joshsamuelson) [inline manifests/params.pp]: So... To do?

Comment (binford2k) [inline manifests/params.pp]: yes, probably a good call. I don't know what the AIO on windows looks like yet

Comment (klynton) [inline manifests/params.pp]: On Windows the codedir is C:\ProgramData\PuppetLabs\code and the confdir stayed the same as before.

**Pull request merged**